### PR TITLE
fix(PubSub): Invalid Arguments error should not be thrown to the client

### PR DESCRIPTION
**Pull request by Rishabh-V**
This fixes the error in Subscriber client caused by toggling Exactly Once delivery on the subscription.  
Status(StatusCode="InvalidArgument", Detail="Some acknowledgement ids in the request were invalid. This could be because the acknowledgement ids have expired or the acknowledgement ids were malformed.")

### CODE REVIEW

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs:1405]:
```suggestion
                        // Check if it's an RpcException with StatusCode of InvalidArgument and Reason of EXACTLY_ONCE_ACKID_FAILURE.
```

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs
Comment (jskeet) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs:1410]: Also if we've got here, then `rpcException` isn't null, so the first `?` in `rpcException?.GetErrorInfo()?.Reason` can be removed. (The second one is still needed.)

Comment (jskeet) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs:1410]: We should also check the domain, really. That's not *vital* as it's very unlikely that anything else would use a reason of EXACTLY_ONCE_ACKID_FAILURE, but it's really the domain+reason that's expected to be unique.

**Review by jskeet — COMMENTED** [review]
Nit on the commit comment - I believe it needs to be "fix" rather than "Fix"; I think conventional commits are case-sensitive. (Certainly we normally use lower-case categories.)

The fix itself looks fine other than the comments I've left, but it would be good to see four tests:

- A non-RpcException (still thrown)
- An RpcException with a different status code (still thrown)
- An RpcException for InvalidArgument with an ErrorInfo specifying a different reason (still thrown)
- An RpcException for InvalidArgument with the right reason/domain (not thrown)

We could add even more tests than that, but let's be pragmatic. However, I do think we should have at least one rethrown and one not-thrown case.

See https://github.com/googleapis/gax-dotnet/blob/main/Google.Api.Gax.Grpc.Tests/RpcExceptionExtensionsTest.cs for sample code that builds an RpcException with an error info.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs
Comment (pradn) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs:1409]: So the ack/modack behavior should be:
1) Retry Unavailable errors for ack/modack (this already happens at the GAPIC/GRPC level, see [here](https://github.com/googleapis/googleapis/blob/480e3fbafbe4eb61c2118e3cb8d9e0d8eb9a9e36/google/pubsub/v1/pubsub_grpc_service_config.json#L221))
2) Ignore all GRPC errors for ack/modack

Please change the code to ignore all GRPC errors for ack/modack in this PR.

The reason we ignore GRPC errors for ack/modack is because they are best-effort RPCs. The user shouldn't see an exception when these happen. There's a subtlety - what if the subscription is not accessible (PermissionDenied) or doesn't exist (NotFound)? Our thinking around this is that it's not that important that the user sees these errors with ack/modack because the StreamingPull stream will fail if these errors occur and then the whole subscriber object will fail. So no need to do extra work here. We may revisit this choice for these two special error codes though. For now, we'll stick to this in the name of uniformity (the Java client has the behavior described here.)

Once we do the full exactly-once delivery work, the whole behavior will be:
1) Retry Unavailable errors for ack/modack at the GAPIC/GRPC level (no change to this behavior)
2) If exactly-once is disabled, ignore all GRPC errors for ack/modack
3) If exactly-once is enabled, bubble up extra info via futures.

**Review by pradn — CHANGES_REQUESTED** [review]

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs
Comment (pradn) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs:45]: You've already fixed this here: https://github.com/googleapis/google-cloud-dotnet/commit/e78afdf41474e38a266a46cb6ae0489ec56a5266

I think you need to rebase the branch on head.

Comment (jskeet) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs:45]: No, that's for SubscriberClient. This is PublisherClient.

Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs]: Thanks Jon. I have excluded PublisherClient from this PR and will raise a new PR for it.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs]: Swallowed all the gRPC errors in Ack/ModAck call. Have tested thoroughly with a client app for an hour. Created new issue #8291 for writing unit tests as it may require substantial time.

**Review by jskeet — APPROVED** [review]
This looks fine to me. @pradn are you happy with non-gRPC exceptions still being propagated?

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs
Comment (pradn) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs]: Ah, gotcha! Thanks both!

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs
Comment (pradn) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs]: Thank you Rishabh, especially for the thorough testing. Please write the unit tests in this PR itself. We usually prefer it that way unless there's a strong reason otherwise. Yes, this issue is high priority, but we can wait for unit tests.

**Comment by pradn** [comment]
I think its fine for non-gRPC exceptions to be propagated, for now at least. We need a wider discussion around the exactly error reporting behavior for acks/modacks. I'll bring it up at the CPS client libs meeting.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:87]: Nit: please move the properties to above the constructor. I'd also remove the blank lines between the properties, and between the three factory methods. (But keep a line between the three factory methods and the properties.)

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:91]: I don't understand what this or `OnModifyAckDeadline` actually mean. I'd suggest either renaming, or adding an XML summary comment. (If you add a comment, then ignore the bit above about removing blank lines.)

Comment (jskeet) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:91]: Reading a bit more, I see what it's doing, but I suspect there may be a cleaner design. We can revisit once this is in though.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:293]: Nit: propogate => propagate

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:300]: Why are you awaiting something you know will throw, rather than just using `throw _ackModifyAckDeadlineAction.Exception`?

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1017]: Unwrap onto the declaration line

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1032]:
```suggestion
            var ackModifyAckDeadlineAction =
                ackOrModifyAck == null ? AckModifyAckDeadlineAction.Data(rpcException, 2, true, true)
                : ackOrModifyAck.Value ? AckModifyAckDeadlineAction.BadAck(rpcException, 2)
                : AckModifyAckDeadlineAction.BadModifyAckDeadline(rpcException, 2);
```

(That's the way we normally format this sort of multiple-conditional operator.)

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1031]: Please reorder these comments to match the order in the code

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1019]: Nit: I'd typically put the `{` on a line on its own here.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1038]: Is this code ever actually going to call ModifyAckDeadline?

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1053]: Again, unwrap...

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:293]: (Here and elsewhere.)

**Review by jskeet — APPROVED** [review]
Mostly formatting nits. It feels a little odd to have the combinatorial tests for different RPCs behaving differently... that may just be due to the way the calls are made though.

I think it's okay to merge this after fixing the formatting, even though it feels odd.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:83]: Done.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:99]: > I don't understand what this or `OnModifyAckDeadline` actually mean. I'd suggest either renaming, or adding an XML summary comment. (If you add a comment, then ignore the bit above about removing blank lines.)
Added the XML comments.

> Reading a bit more, I see what it's doing, but I suspect there may be a cleaner design. We can revisit once this is in though.
Sure Jon.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1063]: Yes Jon. I confirmed again by debugging. We have explicitly added 70 second delay in handler and ackDeadline is set to lower value to simulate the call.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs]: Done.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs]: I was somehow so keen to use `Task.FromException` that `throw` didn't come to my mind. I have now changed it to throw.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs]: Done.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs]: Unit Tests are now added. Please have a look.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (jskeet) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1063]: Aha, right. It might be worth adding a comment to explain that.

**Review by jskeet — APPROVED** [review]
One comment suggest, but other than that, let's ship it.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs
Comment (Rishabh-V) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs:1063]: Added comment in both the tests.

**Review by pradn — APPROVED** [review]
I'm approving the behavior/implementation change. I'll leave it to Jon to review the tests, too.

**Pull request merged**
